### ISSUE TITLE
Demes models cannot have burnin=0 in general

### ISSUE THREAD

**Issue by molpopgen** [molpopgen/fwdpy11#776]
A burn-in of zero causes some event times to be -1, which is invalid.

cc @apragsdale

**Comment by molpopgen** [comment]
This issue is fundamentally about the C++ back-end.  These models *should* be supported.  The question will be about how best to do so.

**Comment by molpopgen** [comment]
@apragsdale -- can you provide an example model of the sort that triggered this issue?  Am I  right in understanding that the only time this can pop up is a demes-generated model with a mass migration affecting the composition of the population at time 0?

**Comment by apragsdale** [comment]
I'm hitting errors with 0 burn in for other cases as well. There might be a fundamental issue with how the demes input is set up. 

```Python
import fwdpy11
import demes

b = demes.Builder()
b.add_deme(
    name="A",
    epochs=[dict(end_time=100, start_size=100), dict(start_size=200)]
)
g = b.resolve()
demog = fwdpy11.discrete_demography.from_demes(g, burnin=0)
```

errors with

```
TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. fwdpy11._fwdpy11._ll_SetDemeSize(when: int, deme: int, new_size: int, resets_growth_rate: bool)

Invoked with: -1, 0, 200, True
```

I think I didn't really consider this burnin=0 case very carefully before. And I think there might just be an off-by-1 error with how I set up the set deme sizes, etc. Essentially, I don't know if in this model, the size is 200 for 100 generations or 101.

**Comment by molpopgen** [comment]
Now that I see the example, I think I see what's going on.  SetDemeSize sets the size for that generation, such that there are that many new offspring.  This "shouldn't" be like using SetMigrationRates to model a pulse migration event, I think.  But this does bring up the issue of how to test this--for even modest N, these off-by-1 errors will be impossible to distinguish from, say, `moments`.

**Comment by molpopgen** [comment]
I should, perhaps, also think more carefully about *when* events are applied within a generation.  Anyhow, there are a few things to ponder here it seems.

**Comment by apragsdale** [comment]
> I should, perhaps, also think more carefully about _when_ events are applied within a generation. Anyhow, there are a few things to ponder here it seems.

Exactly - this is perhaps where I was confused when setting up the demes code (or maybe wasn't confused then, but am now)

**Comment by molpopgen** [comment]
In hindsight, the [docs](https://molpopgen.github.io/fwdpy11/pages/softselection.html) seem vague to me now on that point, even though there's a whole section devoted to it.

**Comment by molpopgen** [comment]
Actually, maybe the docs aren't that vague--this is the key sentence: `That time refers to the birth time of a generation and the events are applied prior to generating offspring, meaning that the events happen to the parents.`

**Comment by molpopgen** [comment]
For example, the following script:

```py
import fwdpy11

rng = fwdpy11.GSLrng(42)
pop = fwdpy11.DiploidPopulation(100, 1.0)
pdict = {
    "nregions": [],
    "sregions": [],
    "recregions": [],
    "rates": (0, 0, 0),
    "demography": None,
    "simlen": 1,
    "gvalue": fwdpy11.Additive(2.0),
}


class Recorder(object):
    def __call__(self, pop, _):
        print(pop.generation, pop.deme_sizes())


s = [fwdpy11.SetDemeSize(10, 0, 200)]
d = fwdpy11.DiscreteDemography(set_deme_sizes=s)
pdict["demography"] = d
pdict["simlen"] = 20
params = fwdpy11.ModelParams(**pdict)
fwdpy11.evolvets(rng, pop, params, 100, recorder=Recorder())
```

Give the following output:

```
1 (array([0], dtype=int32), array([100]))
2 (array([0], dtype=int32), array([100]))
3 (array([0], dtype=int32), array([100]))
4 (array([0], dtype=int32), array([100]))
5 (array([0], dtype=int32), array([100]))
6 (array([0], dtype=int32), array([100]))
7 (array([0], dtype=int32), array([100]))
8 (array([0], dtype=int32), array([100]))
9 (array([0], dtype=int32), array([100]))
10 (array([0], dtype=int32), array([100]))
11 (array([0], dtype=int32), array([200]))
12 (array([0], dtype=int32), array([200]))
13 (array([0], dtype=int32), array([200]))
14 (array([0], dtype=int32), array([200]))
15 (array([0], dtype=int32), array([200]))
16 (array([0], dtype=int32), array([200]))
17 (array([0], dtype=int32), array([200]))
18 (array([0], dtype=int32), array([200]))
19 (array([0], dtype=int32), array([200]))
20 (array([0], dtype=int32), array([200]))
```

**Comment by molpopgen** [comment]
So it seems like the confusion could be what generation is "affected" by the event?

**Comment by molpopgen** [comment]
Notes to self, transcribed from various Slack messages:

* This issue may entirely be on the Python side, due to a off-by-one error for some (all?) demes events.
* In any case, the best way to get at this is to add a load of tests.  Basically, lots of demes models where some event will immediately affect the first offspring generation of the post-burnin part of the model.  We import each model with burnin=0 and burnin > 0 and see what breaks.

**Comment by molpopgen** [comment]
Closed by #822
